### PR TITLE
Collect outdated indirect dependencies

### PR DESCRIPTION
**Pull request by felixfbecker**
This is a shot at collecting outdated indirect dependencies that are locked through a lockfile.
Instead of only looking at the dependencies in composer.json, we look at all packages in composer.lock.
To prevent from suggesting an update to a version that is not actually valid because a dependency has a constraint on it, the reported versions are filtered by all constraints found from other dependencies.

Depends on https://github.com/dependencies-io/actor-php-composer/pull/6

### CODE REVIEW

**Comment by davegaeddert** [comment]
@felixfbecker I went ahead and pushed some changes to your branch. I added a test case so we could validate what it's doing, and had to fix a few things. Can you take a look and let me know what you think? I think this is a fine route to go, would definitely be useful. I do want to make it optional though for now (disabled by default) so that we don't break existing behavior and so that it's more in line with how all the other actors work currently.

Thanks!

File: entrypoint.php [hunk 1]
```diff
@@ -53,11 +55,49 @@ function schema_version($version) {
...
+
+        foreach ($all_packages as $pkg) {
+            if (isset($pkg['require'])) {
+                foreach ($pkg['require'] as $requiredPkg => $constraint) {
+                    // var_dump($constraint);
+                    if ($requiredPkg === $name) $constraints[] = $constraint;
+                }
+            }
+
+            if (isset($pkg['require-dev'])) {
```
Comment (felixfbecker) [inline entrypoint.php]: devDependencies are only installed for the root package, not for dependencies, so unless I'm missing something this shouldn't be considered as a constraint.

Comment (davegaeddert) [reply entrypoint.php]: Ok, I was going to ask about that. That would make sense.

File: entrypoint.php [hunk 2]
```diff
@@ -53,11 +55,49 @@ function schema_version($version) {
...
+                }
+            }
+        }
+
+        if (!empty($constraints)) {
+            $available = array_filter($available, function($version) use($constraints) {
+                // have to compare each version to each constraint individually
+                foreach ($constraints as $constraint) {
+                    // if any version is not satisfied by one of the constraints, it has to be removed
+                    if (!Semver::satisfies($version, $constraint)) return false;
```
Comment (felixfbecker) [inline entrypoint.php]: My initial version used `Semver::satisfiedBy()`, which I thought does what is desired here:

> Return all versions that satisfy given constraints.

Could you explain the reason you changed it?

Comment (davegaeddert) [reply entrypoint.php]: The constraints argument is expected to be a string, not an array. It didn't work. https://github.com/composer/semver/blob/master/src/Semver.php#L49

Comment (felixfbecker) [reply entrypoint.php]: Oh, right, that makes sense. I was fooled by the plural name

Comment (davegaeddert) [reply entrypoint.php]: Me too!

**Comment by felixfbecker** [comment]
LGTM! Any comments on https://github.com/dependencies-io/actor-php-composer/pull/6?

**Comment by davegaeddert** [comment]
I'll take a look at the actor now, at a glance though that seemed like it made sense (much simpler 😉 ). Assuming all is well, I'll probably merge both of these and get a new "beta" version built and send you a PR to get the beta versions enabled on https://github.com/felixfbecker/vscode-php-intellisense so that you can test it for real.

**Comment by felixfbecker** [comment]
Great! I made a PR to make use of this here: https://github.com/felixfbecker/vscode-php-intellisense/pull/165
Did not enable the setting yet though.

**Comment by felixfbecker** [comment]
@davegaeddert are you sure this change is deployed to latest non-beta? After removing the `beta` version tag I don't see transitive dependencies in the output anymore

https://app.dependencies.io/projects/github/felixfbecker/vscode-php-intellisense/builds/27

**Comment by davegaeddert** [comment]
@felixfbecker you're right, the actor got tagged non-beta but the collector got missed somehow. We're trying to automate/streamline the rest of that pipeline right now...

It should be up now.

**Pull request merged**